pkg/backend/httpstate/state.go

```diff
@@ -48,53 +48,79 @@ type tokenSource struct {
 	done     chan bool
 }
 
-func newTokenSource(ctx context.Context, token string, backend *cloudBackend, update client.UpdateIdentifier,
+func newTokenSource(ctx context.Context, initialToken string, backend *cloudBackend, update client.UpdateIdentifier,
 	duration time.Duration) (*tokenSource, error) {
+	requests, done := make(chan tokenRequest), make(chan bool)
+	ts := &tokenSource{requests: requests, done: done}
+	go ts.handleRequests(ctx, initialToken, backend.client, update, duration)
+	return ts, nil
+}
 
-	// Perform an initial lease renewal.
-	newToken, err := backend.client.RenewUpdateLease(ctx, update, token, duration)
-	if err != nil {
-		return nil, err
+func (ts *tokenSource) handleRequests(
+	ctx context.Context,
+	initialToken string,
+	client *client.Client,
+	update client.UpdateIdentifier,
+	duration time.Duration) {
+
+	renewTicker := time.NewTicker(duration / 8)
+	defer renewTicker.Stop()
+
+	state := struct {
+		token   string    // most recently renewed token
+		error   error     // non-nil indicates a terminal error state
+		expires time.Time // assumed expiry of the token
+	}{
+		token:   initialToken,
+		expires: time.Now().Add(duration),
 	}
 
-	requests, done := make(chan tokenRequest), make(chan bool)
-	go func() {
-		// We will renew the lease after 50% of the duration has elapsed to allow more time for retries.
-		ticker := time.NewTicker(duration / 2)
-		defer ticker.Stop()
-
-		for {
-			select {
-			case <-ticker.C:
-				newToken, err = backend.client.RenewUpdateLease(ctx, update, token, duration)
-				// If we get an error from the backend, leave `err` set and surface it during
-				// the next request for a lease token.
-				if err != nil {
-					logging.V(3).Infof("error renewing lease: %v", err)
-					err = fmt.Errorf("renewing lease: %w", err)
-
-					ticker.Stop()
-				} else {
-					token = newToken
-				}
-
-			case c, ok := <-requests:
-				if !ok {
-					close(done)
-					return
-				}
-
-				// err will be non-nil if the last call to RenewUpdateLease failed.
-				resp := tokenResponse{err: err}
-				if err == nil {
-					resp.token = token
-				}
-				c <- resp
-			}
+	renewUpdateLeaseIfStale := func() {
+		if state.error != nil {
+			return
 		}
-	}()
 
-	return &tokenSource{requests: requests, done: done}, nil
+		now := time.Now()
+
+		// We will renew the lease after 50% of the duration
+		// has elapsed to allow time for retries.
+		stale := now.Add(duration / 2).After(state.expires)
+		if !stale {
```

Comment:
```suggestion
		if stale := now.Add(duration / 2).After(state.expires); !stale {
```

Reply:
Prefer variable name as docu on what this is.